app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt

```diff
@@ -168,33 +180,152 @@ class PopupView : FrameLayout {
+                if (codedValueDomain != null) {
+                    setUpSpinner(codedValueDomain, popupField)
+                    popupFieldEditValue.visibility = View.GONE
+                    popupFieldValue.visibility = View.GONE
+                    popupFieldCvdValues.visibility = View.VISIBLE
+                } else {
+                    popupFieldEditValue.inputType = getInputType(popupManager.getFieldType(popupField))
+                    popupFieldEditValue.visibility = View.VISIBLE
+                    popupFieldValue.visibility = View.GONE
+                    //save original colors
+                    val oldColors: ColorStateList = popupFieldLabel.textColors
+                    // here we assign and hold the values of the editable fields, entered by the user
+                    // in popupAttribute.tempValue
+                    popupFieldEditValue.doAfterTextChanged {
+                        if (popupFieldEditValue.hasFocus()) {
 
-                        if (validationError != null) {
-                            val fieldLabelWithValidationError = popupField.label + ": " + validationError.message
-                            popupFieldLabel.text = fieldLabelWithValidationError
-                            popupFieldLabel.setTextColor(Color.RED)
-                        } else {
-                            popupFieldLabel.text = popupField.label
-                            popupFieldLabel.setTextColor(oldColors)
+                            val validationError: ArcGISRuntimeException? = updateValue(
+                                popupField,
+                                popupFieldEditValue.text.toString()
+                            )
+                            if (validationError != null) {
+                                val fieldLabelWithValidationError =
+                                    popupField.label + ": " + validationError.message
+                                popupFieldLabel.text = fieldLabelWithValidationError
+                                popupFieldLabel.setTextColor(Color.RED)
+                            } else {
+                                popupFieldLabel.text = popupField.label
+                                popupFieldLabel.setTextColor(oldColors)
+                            }
                         }
                     }
                 }
             } else {
                 popupFieldEditValue.visibility = View.GONE
+                popupFieldCvdValues.visibility = View.GONE
                 popupFieldValue.visibility = View.VISIBLE
             }
         }
+
+        /**
+         * Sets up spinner for PopupFields that have a CodedValueDomain.
+         */
+        private fun setUpSpinner(codedValueDomain: CodedValueDomain, popupField: PopupField) {
+
+            val codedValuesNames = mutableListOf<String>()
+            for (codedValue in codedValueDomain.codedValues) {
```

Comment:
This for each could be replaced with `codedValueDomain.codedValues.forEach { codedValuesNames.add(it.name) }`

Reply:
Updated.